server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java

```diff
@@ -70,43 +88,103 @@ public AnnotationResponse addAnnotation(String text, EntityType type, String uui
     @ActionEvent(eventType = EventTypes.EVENT_ANNOTATION_REMOVE, eventDescription = "removing an annotation on an entity")
     public AnnotationResponse removeAnnotation(RemoveAnnotationCmd removeAnnotationCmd) {
         String uuid = removeAnnotationCmd.getUuid();
-        if(LOGGER.isDebugEnabled()) {
-            LOGGER.debug("marking annotation removed: " + uuid);
-        }
         AnnotationVO annotation = annotationDao.findByUuid(uuid);
-        annotationDao.remove(annotation.getId());
+        if (isCallingUserAllowedToRemoveAnnotation(annotation)) {
+            if(LOGGER.isDebugEnabled()) {
+                LOGGER.debug("marking annotation removed: " + uuid);
+            }
+            annotationDao.remove(annotation.getId());
+        } else {
+            throw new CloudRuntimeException("Only administrators or entity owner users can delete annotations, cannot remove annotation: " + uuid);
+        }
+
         return createAnnotationResponse(annotation);
     }
 
+    private boolean isCallingUserAllowedToRemoveAnnotation(AnnotationVO annotation) {
+        if (annotation == null) {
+            return false;
+        }
+        if (isCallingUserAdmin()) {
+            return true;
+        }
+        UserVO callingUser = getCallingUserFromContext();
+        String annotationOwnerUuid = annotation.getUserUuid();
+        return annotationOwnerUuid != null && annotationOwnerUuid.equals(callingUser.getUuid());
+    }
+
+    private UserVO getCallingUserFromContext() {
+        CallContext ctx = CallContext.current();
+        long userId = ctx.getCallingUserId();
+        UserVO userVO = userDao.findById(userId);
+        if (userVO == null) {
+            throw new CloudRuntimeException("Cannot find a user with ID " + userId);
+        }
+        return userVO;
+    }
+
+    private boolean isCallingUserAdmin() {
+        UserVO userVO = getCallingUserFromContext();
+        long accountId = userVO.getAccountId();
+        AccountVO accountVO = accountDao.findById(accountId);
+        if (accountVO == null) {
+            throw new CloudRuntimeException("Cannot find account with ID + " + accountId);
+        }
+        Long roleId = accountVO.getRoleId();
+        Role role = roleService.findRole(roleId);
+        if (role == null) {
+            throw new CloudRuntimeException("Cannot find role with ID " + roleId);
+        }
+        return RoleType.Admin.equals(role.getRoleType()) || RoleType.DomainAdmin.equals(role.getRoleType()) ||
+                RoleType.ResourceAdmin.equals(role.getRoleType());
```

Comment:
We could extract these roles to a list (constant) and verify if it contains the specified role.